src/hera/workflows/_meta_mixins.py

```diff
@@ -372,6 +372,12 @@ def __call__(self, *args, **kwargs) -> Union[None, Step, Task]:
                 return Step(template=self, **kwargs)
 
             if isinstance(_context.pieces[-1], DAG):
+                # Add dependencies based on context if not explicitly provided
+                current_task_depends = _context.pieces[-1]._current_task_depends
+                if current_task_depends and "depends" not in kwargs:
+                    kwargs["depends"] = " && ".join(sorted(current_task_depends))
+                current_task_depends.clear()
+
```

Comment:
Not 100% sure.
But this seems more discussions are needed.

e.g. in this pr's example, following code works as expected

```py
        cut_task = cut(CutInput(strings=["hello", "world", "it's", "hera"], cut_after=1))
        join_task = join(JoinInput(strings=cut_task.first_strings, joiner=" "))

        cut_task >> join_task
```

This code can possibly restrict the construction of DAGs.
for example, let's think about following example

```py
A = task_a_fn(SomePydanticClass1(...))
B = task_b_fn(SomePydanticClass2(arg_1=A.return_1, ...)) # B depends on A's output
C = task_c_fn(SomePydanticClass3(...)) # C doesn't depends on A's output.

# But user want to construct DAG as followings.
A >> [B, C]
```

In above case, hera will raise an error. (e.g. already in some task's depends)
https://github.com/argoproj-labs/hera/blob/2eec4c282557823c661dae6b410b470d13a90be5/src/hera/workflows/task.py#L83-L84

Reply:
In your example, you could just write `A >> C` instead of `A >> [B, C]`

However, I agree that this is an existing limitation of the new Pydantic I/O approach which will need solving separately. My preference would be for user-provided dependency edges to not raise duplication errors (like in your example &mdash; it should be fine to be explicit), and to override automatic ones (e.g. if you do `A.on_success(B)`, that should overwrite the existing dependency between A and B). In the meantime, users can:
 - use the old `A.get_parameter` API to avoid adding dependency edges
 - pass in `requires=""` to disable the automatic generation of requires from dependency edges
 - do `B.requires=""` later

Reply:
Sounds great, but for me it looks better to have same/expected behavior even with Pydantic I/O.
User might expect to be able to write `A >> [B, C]`, but this raises an error.
I assume (checked very limited cases) other nodes are also doesn't generate `depends` attribute based on the returns/arguments.

How about separating this feature to another PR, and support this for regardless of the pydantic io?
Maybe @elliotgunton can give more opinions?

Reply:
> User might expect to be able to write A >> [B, C]

I agree - it would be good to write all my dependencies in one place (and as a bonus in one statement) rather than spread out

> user-provided dependency edges to not raise duplication errors

Yes I think it _might_ make sense (not sure about precedent/existing patterns) to make dependency-setting idempotent, and then use the last call (ignoring whether it was automatic or not):

```py
A.on_success(B)
A.on_failure(B)
```
Will end up only with B's `depends: A.Failed`.